token_dispenser/token_dispenser_lambda.py

```diff
@@ -0,0 +1,137 @@
+    pattern = re.compile(r'^[a-zA-Z0-9]{3,32}$')
+    if pattern.match(client_id):
+        return True
+    else:
+        return False
+
+
+def is_minimum_alive_secs_valid(minimum_alive_secs:int or None) -> bool :
+    if minimum_alive_secs is not None and (not minimum_alive_secs.is_integer()):
+        return False
+
+    if ((minimum_alive_secs is not None and minimum_alive_secs.is_integer()) and
+            (minimum_alive_secs > config.MAX_REQUESTED_ALIVE_SECS or minimum_alive_secs < 0)):
+        return False
+    else:
+        return True
+
+
+def handler(event, context):
+    client_id = event.get('client_id', '')
+    minimum_alive_secs: int|None= int(event.get('minimum_alive_secs', config.DEFAULT_TOKEN_MIN_ALIVE_SECS))
+    # client_id must be alphanumeric
+    if not is_client_id_valid(client_id):
+        return {
+            "statusCode": 400,
+            "body": json.dumps({"error": "client_id is required in request. Must be alpha-numeric"})
+        }
+    if not is_minimum_alive_secs_valid(minimum_alive_secs):
+        return {
+            "statusCode": 422,
+            "body": json.dumps({"error": f"minimum_alive_secs if provided, must be numeric "
+                                         f"and smaller or equal than {config.MAX_REQUESTED_ALIVE_SECS} secs"})
+        }
+    # Set the logging level dynamically
+    log_level = getattr(logging, config.LOG_LEVEL, logging.INFO)
+
+    # Reconfigure the logger with the new log level
+    log_adapter = initialize_logger(log_level, client_id=client_id)
+    # This lambda does not burden user with required minimum_alive_sec. If such field is not provided by the caller
+    # it will be set to the default value.
+    if minimum_alive_secs is None:
+        log_adapter.info(f"client passed in empty minimum_alive_secs, using default value as:"
+                         f" {config.DEFAULT_TOKEN_MIN_ALIVE_SECS}")
+        minimum_alive_secs = config.DEFAULT_TOKEN_MIN_ALIVE_SECS
+    # if the token expected to be expired shorter than the expiration time, then
+    # get a new token, save new token to dynamoDB with new TTL
+    # minimum_alive_secs:int|None = int(json.loads(event['body'])['minimum_alive_secs'])
+    log_adapter.debug(f"Context: {context}")
+    log_adapter.info(f'client_id: {client_id}  minimum_alive_secs: {minimum_alive_secs}')
+
+    token_json = None
+    # Retrieve token structure from DynamoDB using client_id
+    try:
+        # When a request comes in, check dynamoDB first, if not found then get new token
+        token_structure_str:str = get_token_by_client_id(client_id)
+        if token_structure_str is not None:
+            token_json = json.loads(token_structure_str)
+        if token_structure_str is not None and satisfy_minimum_alive_secs(token_json, minimum_alive_secs):
+            log_adapter.debug(f"Found token from cache which satisfied minimum allive secs: {minimum_alive_secs}")
+            return token_structure_str
```

Comment:
Can't we just return the dict directly?

https://docs.aws.amazon.com/lambda/latest/dg/python-handler.html#python-handler-return

Reply:
AWS's lambda caller supports having a dict returned directly instead of a string

https://docs.aws.amazon.com/lambda/latest/dg/python-handler.html#python-handler-return